sympy/tensor/array/arrayop.py

```diff
@@ -188,27 +188,49 @@ def derive_by_array(expr, dx):
 
     """
     from sympy.matrices import MatrixBase
+    from sympy.tensor import SparseNDimArray
     array_types = (Iterable, MatrixBase, NDimArray)
-
     if isinstance(dx, array_types):
         dx = ImmutableDenseNDimArray(dx)
         for i in dx:
             if not i._diff_wrt:
                 raise ValueError("cannot derive by this array")
 
     if isinstance(expr, array_types):
-        expr = ImmutableDenseNDimArray(expr)
+        if isinstance(expr, NDimArray):
+            expr = expr.as_immutable()
+        else:
+            expr = ImmutableDenseNDimArray(expr)
+
         if isinstance(dx, array_types):
-            new_array = [[y.diff(x) for y in expr] for x in dx]
+            if isinstance(expr, SparseNDimArray):
+                new_array = _sparse_array_derive_by_array(expr, dx)
+            else:
+                new_array = [[y.diff(x) for y in expr] for x in dx]
             return type(expr)(new_array, dx.shape + expr.shape)
         else:
-            return expr.diff(dx)
```

Comment:
This line was correct. IIRC, this depends on the applyfunc method, so just fix that method.

Reply:
I think it is a little bit compliacted to modify `applyfunc.` My initial idea is to apply the function only to the dictionary of sprase array, in this case, `diff`, it makes sense. But when it comes to some operations which change 0 to other value, e.g. `lambda x: x+1` , every value should be added by one, not only the ones in the dictionary.
So I only modified the function which triggers applyfunc.

Reply:
I am trying with a test like this to make sure 0 is not changed:
```python
if f(0) == 0 and isinstance(self, SparseNDimArray):
        return type(self)({k: f(v) for k, v in my_dictionary.items()},self.shape)
```
But f(0) raised several errors. Do you have any idea to find a bypath for this test?